### ISSUE TITLE
0.52.0 breaks cross compilation of sys dependencies

### ISSUE THREAD

**Issue by marvin-hansen** [bazelbuild/rules_rust#2917]
The new [0.52.0](https://github.com/bazelbuild/rules_rust/releases/tag/0.52.0) release breaks my build completely.

The issue seems to be related to vendored sys-crates not compiling the underlying c deps anymore.
So far I was able to pinpoint the problem to the following crates:
* mimalloc
* lz4-sys

The only viable workaround is to pin rules_rust to any of the previous versions. 

I made a repro of the issue that builds with the previous 0.51 version and virtually all before, but when using the new 0.52 version, it throws the exact same error that I see in my main mono repo.

Link to the repro: 
https://github.com/marvin-hansen/sys-crates-repro/tree/main

**Comment by UebelAndre** [comment]
What error is thrown? Can you add a snippet of that to this ticket?

**Comment by marvin-hansen** [comment]
Here you go.

Sample error. Full log attached.

Depending on which sys crate gets compiled first, the error message may vary, 
but it is always some C import that was not found. 

```bash
cargo:rerun-if-env-changed=CC_SHELL_ESCAPED_FLAGS
CC_SHELL_ESCAPED_FLAGS = None
cargo:warning=liblz4/lib/lz4.c:224:11: fatal error: 'stdlib.h' file not found
cargo:warning=  224 | # include <stdlib.h>   /* malloc, calloc, free */
cargo:warning=      |           ^~~~~~~~~~
cargo:warning=1 error generated.

--stderr:
```

[error-log.txt](https://github.com/user-attachments/files/17260953/error-log.txt)

**Comment by UebelAndre** [comment]
Hmm, I still haven't had the time to dig into this too deeply, but there are a few changes in this release that affect `cargo_build_script`. Since the failure is occurring during execution of one, I'm wondering if https://github.com/bazelbuild/rules_rust/pull/2911 might be suspect? Can you try reverting it and seeing if your issue is fixed? Otherwise bisecting to find the offending commit would be super helpful!

**Comment by marvin-hansen** [comment]
@UebelAndre Yes, the problem was introduced with https://github.com/bazelbuild/rules_rust/pull/2911
I do a revert and fill a PR as fix unless any other workaround comes up.

**Comment by marvin-hansen** [comment]
Please review PR https://github.com/bazelbuild/rules_rust/pull/2921 that resolves the issue.
I ran a test against a fork and it worked so I believe this PR resolves the issue.

Thank you

**Comment by UebelAndre** [comment]
Cc @krasimirgg

**Comment by vitalybuka** [comment]
> Here you go.
> 
> Sample error. Full log attached.
> 
> Depending on which sys crate gets compiled first, the error message may vary, but it is always some C import that was not found.
> 
> ```shell
> cargo:rerun-if-env-changed=CC_SHELL_ESCAPED_FLAGS
> CC_SHELL_ESCAPED_FLAGS = None
> cargo:warning=liblz4/lib/lz4.c:224:11: fatal error: 'stdlib.h' file not found
> cargo:warning=  224 | # include <stdlib.h>   /* malloc, calloc, free */
> cargo:warning=      |           ^~~~~~~~~~
> cargo:warning=1 error generated.
> 
> --stderr:
> ```
> 
> [error-log.txt](https://github.com/user-attachments/files/17260953/error-log.txt)

So if I understand this log correctly,  is_absolute failed on "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk"

Does this dir exists?

If no, then the issue somewhere there, here we exposing that one.

**Comment by vitalybuka** [comment]
Oh, no, it's different, I lost is_absolute on "=" case :facepalm: 
I'll send a PR with a fix.

**Comment by marvin-hansen** [comment]
@vitalybuka Have you tested the fix from https://github.com/bazelbuild/rules_rust/pull/2922?

Because when I ran the latest commit from rules_rust against the repro I shared, 
I still get the exact same error as before...

Would appreciate if this issue gets re-opened. 

Also, I think it would be useful to add some kind of catch-all test to CI to prevent breaking changes to slip through.

I would be open to volunteer to contribute a test that checks if known sys depds cross compile.

**Comment by vitalybuka** [comment]
> @vitalybuka Have you tested the fix from #2922?

Kind of, as I don't know how exactly original error was generated and I have not Darwin.
But I see in the log obviously wrong: "--sysroot=/private/var/tmp/_bazel_marvin/32a31730b3bb4129d41351ed64887532/sandbox/darwin-sandbox/430/execroot/_main//Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk""

If I pass there abs path, I can reproduce similar invalid path as in the log and #2922 fixes it.

Do you still see invalid "sysroot"?

Can you please share the new log?

> 
> Because when I ran the latest commit from rules_rust against the repro I shared, I still get the exact same error as before...
> 
> Would appreciate if this issue gets re-opened.

I don't have permission to reopen.

> 
> Also, I think it would be useful to add some kind of catch-all test to CI to prevent breaking changes to slip through.
> 
> I would be open to volunteer to contribute a test that checks if known sys depds cross compile.

**Comment by vitalybuka** [comment]
> @vitalybuka Have you tested the fix from #2922?
> 
> Because when I ran the latest commit from rules_rust against the repro I shared, I still get the exact same error as before...

I suspect after #2922 
"fix_next_arg = (opt != None)" -> "fix_next_arg = (opt != None and path == None)"
@marvin-hansen Could you try?

@UebelAndre  If not, please revert both  #2922 and #2911

**Comment by marvin-hansen** [comment]
ok,

the issue also occurs on Linux, pick any version you like. 
I was able to reproduce it on Ubuntu 2204.

 The original error was generated with the self not repo I shared, which triggers the error on Mac and Linux:
https://github.com/marvin-hansen/sys-crates-repro/tree/main

That said, If it is possible to just revert https://github.com/bazelbuild/rules_rust/pull/2922 and https://github.com/bazelbuild/rules_rust/pull/2911 this entire issue is set and and done;

Thank you for your time and effort to find a solution.

**Comment by vitalybuka** [comment]
> ok,
> 
> the issue also occurs on Linux, pick any version you like. I was able to reproduce it on Ubuntu 2204.
> 
> The original error was generated with the self-contained repo I shared, which triggers the error on Mac and Linux: https://github.com/marvin-hansen/sys-crates-repro/tree/main

I've created Ubuntu 22.04 VM and can't reproduce:

```
bazel run //test-crate:test-crate
INFO: Analyzed target //test-crate:test-crate (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //test-crate:test-crate up-to-date:
  target-bzl/bin/test-crate/test-crate
INFO: Elapsed time: 0.163s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: target-bzl/bin/test-crate/test-crate
Hello, world!
```

**Comment by marvin-hansen** [comment]
Ok, let me try again with a new VM and fresh clone.

Might have missed something,

**Comment by vitalybuka** [comment]
my relevant history on clean Ubuntu 22.04 VM on GCE
```
    2  sudo apt update
    3  sudo apt install apt-transport-https curl gnupg -y
    4  curl -fsSL https://bazel.build/bazel-release.pub.gpg | gpg --dearmor >bazel-archive-keyring.gpg
    5  sudo mv bazel-archive-keyring.gpg /usr/share/keyrings
    6  echo "deb [arch=amd64 signed-by=/usr/share/keyrings/bazel-archive-keyring.gpg] https://storage.googleapis.com/bazel-apt stable jdk1.8" | sudo tee /etc/apt/sources.list.d/bazel.list
    7  sudo apt update
    8  sudo apt install bazel git
    9  git clone https://github.com/marvin-hansen/sys-crates-repro.git
   10  cd sys-crates-repro/
   12  sudo apt install libtinfo5
   14  bazel build ...
   27  bazel run //test-crate:test-crate
```

**Comment by marvin-hansen** [comment]
@vitalybuka You correct, the repro builds and runs on Ubuntu.
I just tested it on a new Ubuntu VM in Docker. I suspect that the Orbstack machine I used on Friday isn't fully isolated.
Sorry for the confusion. 

On MacOS 15, I still see the same error;
Full log below:
```
bazel build //... --extra_execution_platforms=//build/platforms:darwin-aarch64
INFO: Analyzed 5 targets (0 packages loaded, 0 targets configured).
INFO: Found 5 targets...
ERROR: /Users/marvin/RustroverProjects/forks/sys-crates-repro/thirdparty/crates/libmimalloc-sys-0.1.39/BUILD.bazel:88:19: Running Cargo build script libmimalloc-sys failed: (Exit 1): cargo_build_script_runner failed: error executing CargoBuildScriptRun command (from target //thirdparty/crates/libmimalloc-sys-0.1.39:_bs)
  (cd /private/var/tmp/_bazel_marvin/32a31730b3bb4129d41351ed64887532/sandbox/darwin-sandbox/105/execroot/_main && \
  exec env - \
    AR=external/toolchains_llvm~~llvm~llvm_toolchain/bin/libtool \
    CARGO_CRATE_NAME=libmimalloc_sys \
    CARGO_ENCODED_RUSTFLAGS='--sysroot=${pwd}/bazel-out/darwin_arm64-fastbuild/bin/external/rules_rust~~rust~rust_darwin_aarch64__aarch64-apple-darwin__stable_tools/rust_toolchain\x1f--cap-lints=allow' \
    CARGO_MANIFEST_DIR=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs-.runfiles/_main/thirdparty/crates/libmimalloc-sys-0.1.39 \
    CARGO_MANIFEST_LINKS=mimalloc \
    CARGO_PKG_NAME=libmimalloc-sys \
    CARGO_PKG_VERSION=0.1.39 \
    CARGO_PKG_VERSION_MAJOR=0 \
    CARGO_PKG_VERSION_MINOR=1 \
    CARGO_PKG_VERSION_PATCH=39 \
    CARGO_PKG_VERSION_PRE='' \
    CC=external/toolchains_llvm~~llvm~llvm_toolchain/bin/cc_wrapper.sh \
    CFLAGS='-U_FORTIFY_SOURCE --target=aarch64-apple-macosx -U_FORTIFY_SOURCE -fstack-protector -fno-omit-frame-pointer -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -mmacosx-version-min=15.0 --sysroot=${pwd}//Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk -no-canonical-prefixes -Wno-builtin-macro-redefined -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -fdebug-prefix-map=external/toolchains_llvm~~llvm~llvm_toolchain_llvm/=__bazel_toolchain_llvm_repo__/' \
    CXX=external/toolchains_llvm~~llvm~llvm_toolchain/bin/cc_wrapper.sh \
    CXXFLAGS='-U_FORTIFY_SOURCE --target=aarch64-apple-macosx -U_FORTIFY_SOURCE -fstack-protector -fno-omit-frame-pointer -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -std=c++17 -stdlib=libc++ -Xclang -fno-cxx-modules -mmacosx-version-min=15.0 --sysroot=${pwd}//Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk -no-canonical-prefixes -Wno-builtin-macro-redefined -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -fdebug-prefix-map=external/toolchains_llvm~~llvm~llvm_toolchain_llvm/=__bazel_toolchain_llvm_repo__/' \
    DEBUG=true \
    HOST=aarch64-apple-darwin \
    LD=external/toolchains_llvm~~llvm~llvm_toolchain/bin/cc_wrapper.sh \
    LDFLAGS='-mmacosx-version-min=15.0 -no-canonical-prefixes -fobjc-link-runtime --target=aarch64-apple-macosx -lm -no-canonical-prefixes -headerpad_max_install_names -fobjc-link-runtime -L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk/usr/lib -lc++ -lc++abi -Bdynamic -Lexternal/toolchains_llvm~~llvm~llvm_toolchain_llvm/lib -Bstatic -lunwind -lSystem -lresolv --sysroot=${pwd}//Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk' \
    NUM_JOBS=1 \
    OPT_LEVEL=0 \
    PATH=/Users/marvin/Library/Caches/bazelisk/downloads/sha256/79e4d2401b1c969d2b57babd0c50b0ca61b719689d8b029c919525744f52872c/bin:/Users/marvin/.rd/bin:/Library/Frameworks/Python.framework/Versions/3.12/bin:/Users/marvin/.cargo/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Library/Apple/usr/bin:/usr/local/MacGPG2/bin:/usr/local/go/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Users/marvin/.orbstack/bin \
    PROFILE=debug \
    RUSTC=bazel-out/darwin_arm64-fastbuild/bin/external/rules_rust~~rust~rust_darwin_aarch64__aarch64-apple-darwin__stable_tools/rust_toolchain/bin/rustc \
    TARGET=aarch64-apple-darwin \
    ZERO_AR_DATE=1 \
  bazel-out/darwin-aarch64-opt-exec/bin/external/rules_rust~/cargo/cargo_build_script_runner/cargo_build_script_runner '--script=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs-' '--links=mimalloc' '--out_dir=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.out_dir' '--env_out=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.env' '--flags_out=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.flags' '--link_flags=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.linkflags' '--link_search_paths=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.linksearchpaths' '--dep_env_out=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.depenv' '--stdout=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.stdout.log' '--stderr=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.stderr.log' '--rundir=')
# Configuration: 1c3c1b1160399b2531f3d50669d42c24648dc16670ef4ba3cfeda0c9b36ca211
# Execution platform: //build/platforms:darwin-aarch64

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
Build script process failed with exit code 1
--stdout:
OPT_LEVEL = Some(0)
TARGET = Some(aarch64-apple-darwin)
OUT_DIR = Some(/private/var/tmp/_bazel_marvin/32a31730b3bb4129d41351ed64887532/sandbox/darwin-sandbox/105/execroot/_main/bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.out_dir)
HOST = Some(aarch64-apple-darwin)
cargo:rerun-if-env-changed=CC_aarch64-apple-darwin
CC_aarch64-apple-darwin = None
cargo:rerun-if-env-changed=CC_aarch64_apple_darwin
CC_aarch64_apple_darwin = None
cargo:rerun-if-env-changed=HOST_CC
HOST_CC = None
cargo:rerun-if-env-changed=CC
CC = Some(/private/var/tmp/_bazel_marvin/32a31730b3bb4129d41351ed64887532/sandbox/darwin-sandbox/105/execroot/_main/external/toolchains_llvm~~llvm~llvm_toolchain/bin/cc_wrapper.sh)
RUSTC_WRAPPER = None
cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
CRATE_CC_NO_DEFAULTS = None
DEBUG = Some(true)
cargo:rerun-if-env-changed=MACOSX_DEPLOYMENT_TARGET
MACOSX_DEPLOYMENT_TARGET = None
cargo:rerun-if-env-changed=CFLAGS_aarch64-apple-darwin
CFLAGS_aarch64-apple-darwin = None
cargo:rerun-if-env-changed=CFLAGS_aarch64_apple_darwin
CFLAGS_aarch64_apple_darwin = None
cargo:rerun-if-env-changed=HOST_CFLAGS
HOST_CFLAGS = None
cargo:rerun-if-env-changed=CFLAGS
CFLAGS = Some(-U_FORTIFY_SOURCE --target=aarch64-apple-macosx -U_FORTIFY_SOURCE -fstack-protector -fno-omit-frame-pointer -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -mmacosx-version-min=15.0 --sysroot=/private/var/tmp/_bazel_marvin/32a31730b3bb4129d41351ed64887532/sandbox/darwin-sandbox/105/execroot/_main//Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk -no-canonical-prefixes -Wno-builtin-macro-redefined -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -fdebug-prefix-map=external/toolchains_llvm~~llvm~llvm_toolchain_llvm/=__bazel_toolchain_llvm_repo__/)
cargo:rerun-if-env-changed=CC_SHELL_ESCAPED_FLAGS
CC_SHELL_ESCAPED_FLAGS = None
OUT_DIR = Some(/private/var/tmp/_bazel_marvin/32a31730b3bb4129d41351ed64887532/sandbox/darwin-sandbox/105/execroot/_main/bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.out_dir)
cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
CRATE_CC_NO_DEFAULTS = None
cargo:rerun-if-env-changed=MACOSX_DEPLOYMENT_TARGET
MACOSX_DEPLOYMENT_TARGET = None
HOST = Some(aarch64-apple-darwin)
cargo:rerun-if-env-changed=CFLAGS_aarch64-apple-darwin
CFLAGS_aarch64-apple-darwin = None
cargo:rerun-if-env-changed=CFLAGS_aarch64_apple_darwin
CFLAGS_aarch64_apple_darwin = None
cargo:rerun-if-env-changed=HOST_CFLAGS
HOST_CFLAGS = None
cargo:rerun-if-env-changed=CFLAGS
CFLAGS = Some(-U_FORTIFY_SOURCE --target=aarch64-apple-macosx -U_FORTIFY_SOURCE -fstack-protector -fno-omit-frame-pointer -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -mmacosx-version-min=15.0 --sysroot=/private/var/tmp/_bazel_marvin/32a31730b3bb4129d41351ed64887532/sandbox/darwin-sandbox/105/execroot/_main//Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk -no-canonical-prefixes -Wno-builtin-macro-redefined -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -fdebug-prefix-map=external/toolchains_llvm~~llvm~llvm_toolchain_llvm/=__bazel_toolchain_llvm_repo__/)
cargo:rerun-if-env-changed=CC_SHELL_ESCAPED_FLAGS
CC_SHELL_ESCAPED_FLAGS = None
cargo:warning=In file included from c_src/mimalloc/src/static.c:17:
cargo:warning=c_src/mimalloc/include/mimalloc/internal.h:59:10: fatal error: 'pthread.h' file not found
cargo:warning=#include <pthread.h>
cargo:warning=         ^~~~~~~~~~~
cargo:warning=1 error generated.

--stderr:


error occurred: Command env -u IPHONEOS_DEPLOYMENT_TARGET "/private/var/tmp/_bazel_marvin/32a31730b3bb4129d41351ed64887532/sandbox/darwin-sandbox/105/execroot/_main/external/toolchains_llvm~~llvm~llvm_toolchain/bin/cc_wrapper.sh" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-2" "-fno-omit-frame-pointer" "--target=arm64-apple-darwin" "-mmacosx-version-min=15.0" "-I" "c_src/mimalloc/include" "-I" "c_src/mimalloc/src" "-U_FORTIFY_SOURCE" "--target=aarch64-apple-macosx" "-U_FORTIFY_SOURCE" "-fstack-protector" "-fno-omit-frame-pointer" "-fcolor-diagnostics" "-Wall" "-Wthread-safety" "-Wself-assign" "-mmacosx-version-min=15.0" "--sysroot=/private/var/tmp/_bazel_marvin/32a31730b3bb4129d41351ed64887532/sandbox/darwin-sandbox/105/execroot/_main//Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk" "-no-canonical-prefixes" "-Wno-builtin-macro-redefined" "-D__DATE__=\"redacted\"" "-D__TIMESTAMP__=\"redacted\"" "-D__TIME__=\"redacted\"" "-fdebug-prefix-map=external/toolchains_llvm~~llvm~llvm_toolchain_llvm/=__bazel_toolchain_llvm_repo__/" "-ftls-model=initial-exec" "-DMI_DEBUG=0" "-o" "/private/var/tmp/_bazel_marvin/32a31730b3bb4129d41351ed64887532/sandbox/darwin-sandbox/105/execroot/_main/bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.out_dir/98cfcaec7182b1d8-static.o" "-c" "c_src/mimalloc/src/static.c" with args cc_wrapper.sh did not execute successfully (status code exit status: 1).



ERROR: /Users/marvin/RustroverProjects/forks/sys-crates-repro/thirdparty/crates/libmimalloc-sys-0.1.39/BUILD.bazel:14:13 Compiling Rust rlib libmimalloc_sys v0.1.39 (3 files) failed: (Exit 1): cargo_build_script_runner failed: error executing CargoBuildScriptRun command (from target //thirdparty/crates/libmimalloc-sys-0.1.39:_bs)
  (cd /private/var/tmp/_bazel_marvin/32a31730b3bb4129d41351ed64887532/sandbox/darwin-sandbox/105/execroot/_main && \
  exec env - \
    AR=external/toolchains_llvm~~llvm~llvm_toolchain/bin/libtool \
    CARGO_CRATE_NAME=libmimalloc_sys \
    CARGO_ENCODED_RUSTFLAGS='--sysroot=${pwd}/bazel-out/darwin_arm64-fastbuild/bin/external/rules_rust~~rust~rust_darwin_aarch64__aarch64-apple-darwin__stable_tools/rust_toolchain\x1f--cap-lints=allow' \
    CARGO_MANIFEST_DIR=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs-.runfiles/_main/thirdparty/crates/libmimalloc-sys-0.1.39 \
    CARGO_MANIFEST_LINKS=mimalloc \
    CARGO_PKG_NAME=libmimalloc-sys \
    CARGO_PKG_VERSION=0.1.39 \
    CARGO_PKG_VERSION_MAJOR=0 \
    CARGO_PKG_VERSION_MINOR=1 \
    CARGO_PKG_VERSION_PATCH=39 \
    CARGO_PKG_VERSION_PRE='' \
    CC=external/toolchains_llvm~~llvm~llvm_toolchain/bin/cc_wrapper.sh \
    CFLAGS='-U_FORTIFY_SOURCE --target=aarch64-apple-macosx -U_FORTIFY_SOURCE -fstack-protector -fno-omit-frame-pointer -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -mmacosx-version-min=15.0 --sysroot=${pwd}//Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk -no-canonical-prefixes -Wno-builtin-macro-redefined -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -fdebug-prefix-map=external/toolchains_llvm~~llvm~llvm_toolchain_llvm/=__bazel_toolchain_llvm_repo__/' \
    CXX=external/toolchains_llvm~~llvm~llvm_toolchain/bin/cc_wrapper.sh \
    CXXFLAGS='-U_FORTIFY_SOURCE --target=aarch64-apple-macosx -U_FORTIFY_SOURCE -fstack-protector -fno-omit-frame-pointer -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -std=c++17 -stdlib=libc++ -Xclang -fno-cxx-modules -mmacosx-version-min=15.0 --sysroot=${pwd}//Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk -no-canonical-prefixes -Wno-builtin-macro-redefined -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" -fdebug-prefix-map=external/toolchains_llvm~~llvm~llvm_toolchain_llvm/=__bazel_toolchain_llvm_repo__/' \
    DEBUG=true \
    HOST=aarch64-apple-darwin \
    LD=external/toolchains_llvm~~llvm~llvm_toolchain/bin/cc_wrapper.sh \
    LDFLAGS='-mmacosx-version-min=15.0 -no-canonical-prefixes -fobjc-link-runtime --target=aarch64-apple-macosx -lm -no-canonical-prefixes -headerpad_max_install_names -fobjc-link-runtime -L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk/usr/lib -lc++ -lc++abi -Bdynamic -Lexternal/toolchains_llvm~~llvm~llvm_toolchain_llvm/lib -Bstatic -lunwind -lSystem -lresolv --sysroot=${pwd}//Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.0.sdk' \
    NUM_JOBS=1 \
    OPT_LEVEL=0 \
    PATH=/Users/marvin/Library/Caches/bazelisk/downloads/sha256/79e4d2401b1c969d2b57babd0c50b0ca61b719689d8b029c919525744f52872c/bin:/Users/marvin/.rd/bin:/Library/Frameworks/Python.framework/Versions/3.12/bin:/Users/marvin/.cargo/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Library/Apple/usr/bin:/usr/local/MacGPG2/bin:/usr/local/go/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Users/marvin/.orbstack/bin \
    PROFILE=debug \
    RUSTC=bazel-out/darwin_arm64-fastbuild/bin/external/rules_rust~~rust~rust_darwin_aarch64__aarch64-apple-darwin__stable_tools/rust_toolchain/bin/rustc \
    TARGET=aarch64-apple-darwin \
    ZERO_AR_DATE=1 \
  bazel-out/darwin-aarch64-opt-exec/bin/external/rules_rust~/cargo/cargo_build_script_runner/cargo_build_script_runner '--script=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs-' '--links=mimalloc' '--out_dir=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.out_dir' '--env_out=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.env' '--flags_out=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.flags' '--link_flags=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.linkflags' '--link_search_paths=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.linksearchpaths' '--dep_env_out=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.depenv' '--stdout=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.stdout.log' '--stderr=bazel-out/darwin_arm64-fastbuild/bin/thirdparty/crates/libmimalloc-sys-0.1.39/_bs.stderr.log' '--rundir=')
# Configuration: 1c3c1b1160399b2531f3d50669d42c24648dc16670ef4ba3cfeda0c9b36ca211
# Execution platform: //build/platforms:darwin-aarch64

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
INFO: Elapsed time: 2.986s, Critical Path: 2.76s
INFO: 9 processes: 9 internal.
ERROR: Build did NOT complete successfully
```

**Comment by krasimirgg** [comment]
I wonder if we could turn the example from https://github.com/marvin-hansen/sys-crates-repro/tree/main into an regression "examples" test. We have for example this CI config that builds the crate universe example on macos platform: https://github.com/bazelbuild/rules_rust/blob/2a85e13c74239d2a0d0c47a37f10a4c9739fa5e6/.bazelci/presubmit.yml#L532

**Comment by marvin-hansen** [comment]
Yes, we can do that!

Essentially, I could update the the cross compilation example in the
bazelmod folder, wire it in the CI and let it run whenever a new mergeque
comes in.

I can fill a PR for that if you want.

LMK
M

**Comment by krasimirgg** [comment]
> Yes, we can do that! Essentially, I could update the the cross compilation example in the bazelmod folder, wire it in the CI and let it run whenever a new mergeque comes in. I can fill a PR for that if you want. LMK M

Thank you! Yes please!

**Comment by marvin-hansen** [comment]
Alright,

You should get something to review by tomorrow when I'm back in office.

**Comment by vitalybuka** [comment]
> Alright, You should get something to review by tomorrow when I'm back in office.

@krasimirgg Meanwhile should reopen the issue and fix forward with https://github.com/bazelbuild/rules_rust/pull/2924 (or revert) ?

**Comment by UebelAndre** [comment]
I've not caught up with most of the activity in this thread but I have a revert PR here: https://github.com/bazelbuild/rules_rust/pull/2925

For whatever change lands next to either restore the functionality or fix forward, I think it should have a regression test. If it's possible to have a small test (that doesn't rely on external crates) at `//test/cargo_build_script` that'd be ideal. Using `//examples` as regression tests is convenient but it's not clear what specifically is being tested and if there's ever a need to refactor an example it seems likely that some regression tests would be dropped.

**Comment by vitalybuka** [comment]
> I've not caught up with most of the activity in this thread but I have a revert PR here: #2925
> 
> For whatever change lands next to either restore the functionality or fix forward, I think it should have a regression test. If it's possible to have a small test (that doesn't rely on external crates) at `//test/cargo_build_script` that'd be ideal. Using `//examples` as regression tests is convenient but it's not clear what specifically is being tested and if there's ever a need to refactor an example it seems likely that some regression tests would be dropped.

My problem with the test that we need to test `_pwd_flags` which receives args from cc_toolchain features.
and rules_rust/ does not define cc_toolchain.
Is my understanding correct?

**Comment by marvin-hansen** [comment]
I updated the cross compilation example to catch potential issues with cross compiling sys dependencies.
Please review PR: https://github.com/bazelbuild/rules_rust/pull/2926

Also, I want to clarify that this is just a stopgap for catching issues related to cross compiling sys dependencies and does not replace proper testing of the rules itself. AFAIK, it is not fully understood yet why https://github.com/bazelbuild/rules_rust/pull/2922 causes the issue on MacOS so the big idea is to use the updated example as a proxy until the real cause has been identified and proper regression testing has been added. IMHO, it is the best we can do right now to prevent breaking changes to slip through again. I acknowledge its not great, but please feel free to implement an improved mechanism. 

I want to thank everyone involved for all the effort and dedication to resolve the issue. 

Since the test passes (because the underlying issue has been reverted), I suggest closing this issue as the next release should rollout the fixture.

Thank you

**Comment by krasimirgg** [comment]
> > I've not caught up with most of the activity in this thread but I have a revert PR here: #2925
> > For whatever change lands next to either restore the functionality or fix forward, I think it should have a regression test. If it's possible to have a small test (that doesn't rely on external crates) at `//test/cargo_build_script` that'd be ideal. Using `//examples` as regression tests is convenient but it's not clear what specifically is being tested and if there's ever a need to refactor an example it seems likely that some regression tests would be dropped.
> 
> My problem with the test that we need to test `_pwd_flags` which receives args from cc_toolchain features. and rules_rust/ does not define cc_toolchain. Is my understanding correct?

I'm looking into how we can control this stuff in unit test context. It's a bit involved, but I hope I have something to share tomorrow.

**Comment by marvin-hansen** [comment]
I think this issue can be closed as the latest 0.53 releases added some regression tests.
